service/history/queues/queue_base.go

```diff
@@ -0,0 +1,356 @@
+			QueueTypeUnknown, // we don't care about queue type
+			shard.GetConfig().NamespaceCacheRefreshInterval,
+		)
+	}
+
+	readerInitializer := func(readerID int32, slices []Slice) Reader {
+		return NewReader(
+			slices,
+			&options.ReaderOptions,
+			scheduler,
+			rescheduler,
+			timeSource,
+			rateLimiter,
+			logger,
+			metricsHandler,
+		)
+	}
+
+	inclusiveLowWatermark := exclusiveHighWatermark
+	readerGroup := NewReaderGroup(readerInitializer)
+	for readerID, scopes := range readerScopes {
+		slices := make([]Slice, 0, len(scopes))
+		for _, scope := range scopes {
+			slices = append(slices, NewSlice(paginationFnProvider, executableInitializer, scope))
+		}
+		readerGroup.NewReader(readerID, slices...)
+
+		if len(scopes) != 0 {
+			inclusiveLowWatermark = tasks.MinKey(inclusiveLowWatermark, scopes[0].Range.InclusiveMin)
+		}
+	}
+
+	return &queueBase{
+		shard: shard,
+
+		status:     common.DaemonStatusInitialized,
+		shutdownCh: make(chan struct{}),
+
+		category:    category,
+		options:     options,
+		rescheduler: rescheduler,
+		timeSource:  shard.GetTimeSource(),
+		logger:      logger,
+
+		paginationFnProvider:  paginationFnProvider,
+		executableInitializer: executableInitializer,
+
+		inclusiveLowWatermark: inclusiveLowWatermark,
+		nonReadableScope: NewScope(
+			NewRange(exclusiveHighWatermark, tasks.MaximumKey),
+			predicates.Universal[tasks.Task](),
+		),
+		readerGroup: readerGroup,
+
+		// pollTimer and checkpointTimer are initialized on Start()
+		checkpointRetrier: backoff.NewRetrier(
+			createCheckpointRetryPolicy(),
+			backoff.SystemClock,
+		),
+	}
+}
+
+func (p *queueBase) Start() {
+	p.rescheduler.Start()
+	p.readerGroup.Start()
+
+	p.pollTimer = time.NewTimer(backoff.JitDuration(
+		p.options.MaxPollInterval(),
+		p.options.MaxPollIntervalJitterCoefficient(),
+	))
+	p.checkpointTimer = time.NewTimer(backoff.JitDuration(
+		p.options.CheckpointInterval(),
+		p.options.CheckpointIntervalJitterCoefficient(),
+	))
+}
+
+func (p *queueBase) Stop() {
+	p.readerGroup.Stop()
+	p.rescheduler.Stop()
+	p.pollTimer.Stop()
+	p.checkpointTimer.Stop()
+}
+
+func (p *queueBase) Category() tasks.Category {
+	return p.category
+}
+
+func (p *queueBase) FailoverNamespace(
+	namespaceIDs map[string]struct{},
+) {
+	// TODO: reschedule all tasks for namespaces that becomes active
+	// no-op
+}
+
+func (p *queueBase) LockTaskProcessing() {
+	// no-op
+}
+
+func (p *queueBase) UnlockTaskProcessing() {
+	// no-op
+}
+
+func (p *queueBase) processPollTimer() {
+	if p.lastPollTime.Add(p.options.MaxPollInterval()).Before(p.timeSource.Now()) {
+		p.processNewRange()
+	}
+
+	p.pollTimer.Reset(backoff.JitDuration(
+		p.options.MaxPollInterval(),
+		p.options.MaxPollIntervalJitterCoefficient(),
+	))
+}
+
+func (p *queueBase) processNewRange() {
+	newMaxKey := p.shard.GetQueueExclusiveHighReadWatermark(
+		p.category,
+		p.shard.GetClusterMetadata().GetCurrentClusterName(),
+	)
+
+	if !p.nonReadableScope.CanSplitByRange(newMaxKey) {
```

Comment:
is the non readable scope the same as the write scope?

Reply:
Yes, logically they are the same. But the non-readable scope here will be sightly behind the one in shard info, as the boundary is updated later.